File: lib/tables.py
"""
Simple Pydantic models corresponding to database tables
"""

from datetime import datetime
from sys import last_type
from typing import Optional, List, Dict, Any
from uuid import UUID
from decimal import Decimal
from pydantic import BaseModel


class UserProfile(BaseModel):
    id: UUID
    display_name: Optional[str] = None
    deepgram_enabled: bool
    base_language_model: str
    general_instructions: str    
    wake_word: str = "Jarvis"
    wake_word_detection_enabled: bool = True
    wake_word_sensitivity: float = 0.8
    selected_deepgram_voice: str = "aura-2-mars-en"
    timezone: str = "UTC"
    preferences: Dict[str, Any] = {}
    # XAI LiveSearch settings
    xai_live_search_enabled: Optional[bool] = False
    xai_live_search_safe_search: Optional[bool] = True
    created_at: datetime
    updated_at: Optional[datetime] = None


class Conversation(BaseModel):
    id: UUID
    user_id: UUID
    title: Optional[str] = None
    summary: Optional[str] = None
    conversation_type: str = "general"
    status: str = "active"
    metadata: Dict[str, Any] = {}
    created_at: datetime
    updated_at: datetime


class Message(BaseModel):
    id: UUID
    conversation_id: UUID
    user_id: UUID
    role: str
    content: str
    audio_url: Optional[str] = None
    transcription_confidence: Optional[float] = None
    tool_calls: Optional[Dict[str, Any]] = None
    metadata: Dict[str, Any] = {}
    created_at: datetime


class Memory(BaseModel):
    id: UUID
    user_id: UUID
    memory_type: str
    category: Optional[str] = None
    title: str
    content: str
    tags: List[str] = []
    importance_score: int = 1
    embedding: Optional[List[float]] = None
    decay_factor: float = 1.0
    auto_committed: bool = False
    source_conversation_id: Optional[UUID] = None
    last_accessed: datetime
    created_at: datetime
    updated_at: datetime


class UserHabit(BaseModel):
    id: UUID
    user_id: UUID
    habit_type: str
    pattern: str
    frequency_data: Dict[str, Any] = {}
    confidence_score: float = 0.5
    suggested_automation: Optional[str] = None
    automation_approved: bool = False
    last_observed: datetime
    created_at: datetime


class Automation(BaseModel):
    id: UUID
    user_id: UUID
    name: str
    trigger_conditions: Dict[str, Any]
    actions: Dict[str, Any]
    is_active: bool = True
    execution_count: int = 0
    last_executed: Optional[datetime] = None
    notes: Optional[str] = None
    created_at: datetime


class ConfigForm(BaseModel):
    id: UUID
    service_id: UUID
    json: Dict[str, Any]
    link: Optional[str] = None
    updated_at: datetime


class Integration(BaseModel):
    id: UUID
    user_id: UUID
    service_id: UUID  # points to service id instead of service_name
    notes: Optional[str] = None
    is_active: bool = True
    status: Optional[str] = 'active'  # pending, in_progress, active, inactive, failed
    last_used: Optional[datetime] = None
    created_at: datetime
    # OAuth fields (for calendar and email integrations)
    access_token: Optional[str] = None
    refresh_token: Optional[str] = None
    expires_at: Optional[datetime] = None
    # Email specific fields
    email_address: Optional[str] = None
    # Notion specific fields
    bot_id: Optional[str] = None
    workspace_name: Optional[str] = None
    workspace_icon: Optional[str] = None
    workspace_id: Optional[str] = None
    owner_info: Optional[Dict[str, Any]] = None
    duplicated_template_id: Optional[str] = None
    # Common sync fields
    last_sync: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    client_id: Optional[str] = None
    client_secret_id: Optional[str] = None
    client_secret_value: Optional[str] = None


class Service(BaseModel):
    id: UUID
    created_at: datetime
    service_name: str
    num_users: int = 0
    config_form_json: Optional[Dict[str, Any]] = None  # Cached config form data
    auth_script: Optional[str] = None
    tools: Optional[List[str]] = None
    client_creation_script: Optional[str] = None
    integration_method: Optional[str] = None  # e.g. OAuth, External App, Internal App etc.


class Action(BaseModel):
    id: UUID
    service_id: UUID
    name: str
    description: str
    parameters: Dict[str, Any]  # JSON schema for input parameters
    returns: Dict[str, Any]     # JSON schema for output format
    example: Dict[str, Any]     # Example usage
    run_script: str             # Executable Python script/logic
    endpoint_url: Optional[str] = None  # API endpoint if applicable
    http_method: Optional[str] = None   # GET, POST, etc.
    auth_required: bool = True          # Whether authentication is needed
    category: Optional[str] = None      # e.g., "communication", "storage", "analytics"
    version: str = "1.0"               # Tool version
    is_active: bool = True             # Whether tool is available for use
    execution_timeout: int = 30        # Timeout in seconds
    rate_limit: Optional[int] = None   # Max executions per minute
    created_at: datetime
    updated_at: Optional[datetime] = None


class CancellationRequest(BaseModel):
    id: UUID
    user_id: UUID
    request_id: str             # Unique identifier for the request to cancel
    request_type: str = "chat"  # Type of request (chat, integration, etc.)
    status: str = "pending"     # pending, processed, expired
    metadata: Dict[str, Any] = {}  # Additional cancellation context
    created_at: datetime
    processed_at: Optional[datetime] = None


class IntegrationBuildState(BaseModel):
    """
    Tracks the state and progress of integration building workflow.
    Added in Phase 5 for flow validation and step enforcement.
    """
    id: UUID
    user_id: UUID
    service_name: str
    completed_steps: List[str] = []  # List of completed step names
    current_status: str = "not_started"  # not_started, in_progress, form_ready, auth_ready, completed, failed
    state_data: Dict[str, Any] = {}  # Additional state information
    created_at: datetime
    last_updated: datetime
    updated_by: Optional[UUID] = None

File: lib/tables.ts
export type UserProfile = {
    id: string;
    display_name?: string;
    deepgram_enabled: boolean;
    base_language_model: string;
    general_instructions: string;
    wake_word: string;
    wake_word_sensitivity: number;
    wake_word_detection_enabled: boolean;
    selected_deepgram_voice: string;
    timezone: string;
    preferences: Record<string, any>;
    // XAI LiveSearch settings
    xai_live_search_enabled?: boolean;
    xai_live_search_safe_search?: boolean;
    created_at: Date;
    updated_at: Date;
  };
  
  export const userProfileFields = [
    'id', 'display_name', 'deepgram_enabled', 'base_language_model', 'general_instructions',
    'wake_word', 'wake_word_sensitivity', 'wake_word_detection_enabled', 'selected_deepgram_voice', 'timezone', 'preferences', 
    'xai_live_search_enabled', 'xai_live_search_safe_search',
    'created_at', 'updated_at'
  ] as const;
  export type UserProfileField = (typeof userProfileFields)[number];
  
  export type Conversation = {
    id: string;
    user_id: string;
    title?: string;
    summary?: string;
    conversation_type: string;
    status: string;
    metadata: Record<string, any>;
    created_at: Date;
    updated_at: Date;
  };
  
  export const conversationFields = [
    'id', 'user_id', 'title', 'summary', 'conversation_type', 
    'status', 'metadata', 'created_at', 'updated_at'
  ] as const;
  export type ConversationField = (typeof conversationFields)[number];
  
  export type Message = {
    id: string;
    conversation_id: string;
    user_id: string;
    role: string;
    content: string;
    audio_url?: string;
    transcription_confidence?: number;
    tool_calls?: Record<string, any>;
    metadata: Record<string, any>;
    created_at: Date;
  };
  
  export const messageFields = [
    'id', 'conversation_id', 'user_id', 'role', 'content', 'audio_url',
    'transcription_confidence', 'tool_calls', 'metadata', 'created_at'
  ] as const;
  export type MessageField = (typeof messageFields)[number];
  
  export type Memory = {
    id: string;
    user_id: string;
    memory_type: string;
    category?: string;
    title: string;
    content: string;
    tags: string[];
    importance_score: number;
    embedding?: number[];
    decay_factor: number;
    auto_committed: boolean;
    source_conversation_id?: string;
    last_accessed: Date;
    created_at: Date;
    updated_at: Date;
  };
  
  export const memoryFields = [
    'id', 'user_id', 'memory_type', 'category', 'title', 'content',
    'importance_score', 'embedding', 'decay_factor', 'auto_committed',
    'source_conversation_id', 'last_accessed', 'created_at', 'updated_at'
  ] as const;
  export type MemoryField = (typeof memoryFields)[number];

  
  
  export type UserHabit = {
    id: string;
    user_id: string;
    habit_type: string;
    pattern: string;
    frequency_data: Record<string, any>;
    confidence_score: number;
    suggested_automation?: string;
    automation_approved: boolean;
    last_observed: Date;
    created_at: Date;
  };
  
  export const userHabitFields = [
    'id', 'user_id', 'habit_type', 'pattern', 'frequency_data',
    'confidence_score', 'suggested_automation', 'automation_approved',
    'last_observed', 'created_at'
  ] as const;
  export type UserHabitField = (typeof userHabitFields)[number];
  
  export type Automation = {
    id: string;
    user_id: string;
    name: string;
    trigger_conditions: Record<string, any>;
    actions: Record<string, any>;
    is_active: boolean;
    execution_count: number;
    last_executed?: Date;
    notes?: string;
    created_at: Date;
  };
  
  export const automationFields = [
    'id', 'user_id', 'name', 'trigger_conditions', 'actions',
    'is_active', 'execution_count', 'last_executed', 'notes', 'created_at'
  ] as const;
  export type AutomationField = (typeof automationFields)[number];
  
  
  export type ConfigForm = {
    id: string;
    service_id: string;
    json: Record<string, any>;
    setup_instructions?: string;
    updated_at: Date;
  };

  export const configFormFields = [
    'id', 'service_id', 'json', 'setup_instructions', 'updated_at'
  ] as const;
  export type ConfigFormField = (typeof configFormFields)[number];
  
  export type Integration = {
    id: string;
    user_id: string;
    service_id: string; // points to service id instead of service_name
    notes?: string;
    is_active: boolean;
    status?: string; // pending, in_progress, active, inactive, failed
    last_used?: Date;
    created_at: Date;
    // OAuth fields (for calendar and email integrations)
    access_token?: string;
    refresh_token?: string;
    expires_at?: Date;
    // Email specific fields
    email_address?: string;
    // Notion specific fields
    bot_id?: string;
    workspace_name?: string;
    workspace_icon?: string;
    workspace_id?: string;
    owner_info?: Record<string, any>;
    duplicated_template_id?: string;
    // Common sync fields
    last_sync?: Date;
    updated_at?: Date;
    client_id?: string;
    client_secret_id?: string;
    client_secret_value?: string;
  };
  
  export const integrationFields = [
    'id', 'user_id', 'service_id', 'notes', 'is_active', 'status', 'last_used', 'created_at',
    'access_token', 'refresh_token', 'expires_at', 'email_address',
    'bot_id', 'workspace_name', 'workspace_icon', 'workspace_id', 'owner_info', 'duplicated_template_id',
    'last_sync', 'updated_at', 'client_id', 'client_secret_id', 'client_secret_value'
  ] as const;
  export type IntegrationField = (typeof integrationFields)[number];

  export type Service = {
    id: string;
    created_at: Date;
    service_name: string;
    num_users: number;
    config_form_id?: string; // Foreign key to ConfigForm.id
    auth_script?: string;
    tools?: string[];
    integration_method?: string; // e.g. OAuth, External App, Internal App etc.
  };

  export const serviceFields = [
    'id', 'created_at', 'service_name', 'num_users', 'config_form_id',
    'auth_script', 'tools', 'integration_method'
  ] as const;
  export type ServiceField = (typeof serviceFields)[number];

  export type Action = {
    id: string;
    service_id: string;
    name: string;
    description: string;
    parameters: Record<string, any>;  // JSON schema for input parameters
    returns: Record<string, any>;     // JSON schema for output format
    example: Record<string, any>;     // Example usage
    run_script: string;               // Executable Python script/logic
    endpoint_url?: string;            // API endpoint if applicable
    http_method?: string;             // GET, POST, etc.
    auth_required: boolean;           // Whether authentication is needed
    category?: string;                // e.g., "communication", "storage", "analytics"
    version: string;                  // Tool version
    is_active: boolean;               // Whether tool is available for use
    execution_timeout: number;        // Timeout in seconds
    rate_limit?: number;              // Max executions per minute
    created_at: Date;
    updated_at?: Date;
  };

  export const actionFields = [
    'id', 'service_id', 'name', 'description', 'parameters',
    'returns', 'example', 'run_script', 'endpoint_url', 'http_method',
    'auth_required', 'category', 'version', 'is_active', 'execution_timeout',
    'rate_limit', 'created_at', 'updated_at'
  ] as const;
  export type ActionField = (typeof actionFields)[number];

  export type CancellationRequest = {
    id: string;
    user_id: string;
    request_id: string;          // Unique identifier for the request to cancel
    request_type: string;        // Type of request (chat, integration, etc.)
    status: string;              // pending, processed, expired
    metadata: Record<string, any>; // Additional cancellation context
    created_at: Date;
    processed_at?: Date;
  };

  export const cancellationRequestFields = [
    'id', 'user_id', 'request_id', 'request_type', 'status',
    'metadata', 'created_at', 'processed_at'
  ] as const;
  export type CancellationRequestField = (typeof cancellationRequestFields)[number];

  export type IntegrationBuildState = {
    id: string;
    user_id: string;
    service_name: string;
    completed_steps: string[];      // List of completed step names
    current_status: string;         // not_started, in_progress, form_ready, auth_ready, completed, failed
    state_data: Record<string, any>; // Additional state information
    created_at: Date;
    last_updated: Date;
    updated_by?: string;
  };

  export const integrationBuildStateFields = [
    'id', 'user_id', 'service_name', 'completed_steps', 'current_status',
    'state_data', 'created_at', 'last_updated', 'updated_by'
  ] as const;
  export type IntegrationBuildStateField = (typeof integrationBuildStateFields)[number];



